### DISCUSSION TITLE
Table function for views

### DISCUSSION THREAD

**Discussion by rkennedy-mode** [duckdb/duckdb#8657] [Q&A]
I'm trying to create a bit of a virtual catalog to satisfy queries of the form:

```sql
SELECT location, sum(total_comp)
FROM people_ops.salary
GROUP BY location
```

In this case, the table referenced above is a dynamically generated view, based on a flat file (e.g. CSV or Parquet). The SQL for that view would look something like:

```sql
SELECT
    id,
    first_name,
    last_name,
    location,
    salary,
    bonus,
    salary + bonus as total_comp
FROM 's3://bucket_name/path/to/salary_data.csv'
```

In the above, the `id`, `first_name`, `last_name`, `location`, `salary`, and `bonus` columns are physical colums from the CSV file, while `total_comp` is calculated on the fly.

I'd like to be able to create a [replacement scan](https://duckdb.org/docs/api/c/replacement_scans) for the `people_ops.salary` table, which returns a table function for the view mentioned above. I didn't see a built-in way of doing that, however. So I wondered if there was, maybe, something lurking that I hadn't noticed or maybe a relatively easy way to reproduce the functionality I'm looking for.

**Comment by Tishj** [comment]
I might be misunderstanding, but that sounds like you want a table macro?
<https://duckdb.org/docs/sql/statements/create_macro>

**Reply by rkennedy-mode** [reply]
Not quite. Part of the problem is I can't anticipate which tables will be requested in the SQL query (it comes from an outside party). I'm trying to avoid pre-populating DuckDB with all possible tables as the catalog could be quite large. So using a replacement scan would allow me to only pull in the definitions (views) needed to satisfy the query.

To accomplish the same thing with a macro or even just creating views, I'd need to be able to analyze the query to see which (virtual) tables it needs to access before issuing the query. I don't think there's a practical way of doing that with DuckDB, though I'd be happy to be proven wrong.
